### PR TITLE
Fixed the object mask to work with result limit.

### PR DESCRIPTION
**Pull request by FernandoOjeda**
https://github.com/softlayer/softlayer-java/issues/52

### CODE REVIEW

File: src/main/java/com/softlayer/api/Mask.java [hunk 1]
```diff
@@ -44,7 +45,26 @@ protected String getMask() {
 
     @Override
     public String toString() {
-        return toString(new StringBuilder()).toString();
+        String objectMask = new String();
+        String builtedMask = toString(new StringBuilder()).toString();
```
Comment (allmightyspiff) [inline src/main/java/com/softlayer/api/Mask.java]: should be "builtMask"

File: src/main/java/com/softlayer/api/Mask.java [hunk 2]
```diff
@@ -44,7 +45,26 @@ protected String getMask() {
...
+        if(builtedMask.contains("[")){
```
Comment (allmightyspiff) [inline src/main/java/com/softlayer/api/Mask.java]: I don't really understand what this code is trying to do, can you provide an example Mask that would trigger the `if(builtedMask.contains("[")){` branch and the `else` branch?

Comment (FernandoOjeda) [reply src/main/java/com/softlayer/api/Mask.java]: I´m using this example code:

Account.Service accountService = Account.service(client);

accountService.withMask().blockDeviceTemplateGroups().id();
accountService.withMask().blockDeviceTemplateGroups().blockDevices().diskImage().type();
accountService.withMask().blockDeviceTemplateGroups().status().keyName();

List<Group> result = accountService.getBlockDeviceTemplateGroups();

For the conditional if(builtedMask.contains("["))is when I use more than one mask like above objectMask example.

The method toString() of the class Mask return the value like this:

blockDeviceTemplateGroups[id, blockDevices.diskImage.type, status.keyName]

with this data the objectMask will be building to send the request, when the request is sending like this without to put the getObject option between the accountService and  getBlockDeviceTemplateGroups() method the response will be that the blockDeviceTemplateGroups is not recognized as a value for the objectMask and the service. It is for that I implemented this code to remove the value blockDeviceTemplateGroups to send the correct objectMask without using getObject for the method getBlockDeviceTemplateGroups().

The result for the conditional if() will be:

id, blockDevices.diskImage.type, status.keyName

For the else branch is for one mask, like this example:

accountService.withMask().blockDeviceTemplateGroups().status().keyName();

The method toString() will return like this:

blockDeviceTemplateGroups.status.keyName

And with this branch else the result will be:

status.keyName

Comment (allmightyspiff) [reply src/main/java/com/softlayer/api/Mask.java]: Thanks, that helps me understand what is going on a lot better.

Comment (FernandoOjeda) [reply src/main/java/com/softlayer/api/Mask.java]: I'm glad to hear that, now we can do the request using objectMask and resultLimit at the same time.

**Comment by allmightyspiff** [comment]
The build has an error from the unit tests.
https://travis-ci.org/softlayer/softlayer-java/jobs/390882491#L1856

```
testMaskRemoval(com.softlayer.api.RestApiClientTest)  Time elapsed: 0.008 sec  <<< ERROR!
java.lang.StringIndexOutOfBoundsException: String index out of range: -1
	at java.lang.String.substring(String.java:1967)
	at com.softlayer.api.Mask.toString(Mask.java:66)
	at com.softlayer.api.RestApiClientTest.testMaskRemoval(RestApiClientTest.java:476)
```

File: src/main/java/com/softlayer/api/Mask.java [hunk 3]
```diff
@@ -44,7 +45,26 @@ protected String getMask() {
...
+            String [] mask = builtedMask.split(Pattern.quote("["));
+            for (int count = 0; count < mask.length; count ++ ) {
+                if (count != 0){
```
Comment (allmightyspiff) [inline src/main/java/com/softlayer/api/Mask.java]: You could change the for loop to just start at 1 instead of 0, and remove this if statement

**Comment by allmightyspiff** [comment]
After digging into the code a bit, I don't think fixing this problem in toString() is going to be the right approach.


```java
        accountService.withMask().abuseEmail();
        accountService.withMask().virtualGuests();
	    accountService.withMask().blockDeviceTemplateGroups().id();
	    accountService.withMask().blockDeviceTemplateGroups().name();
        System.out.print("mask[" + accountService.withMask().toString() + "]\n");
	    for (Group bd : accountService.getObject().getBlockDeviceTemplateGroups()) {
            System.out.println("BD: " + bd.getId() + " - " + bd.getName());
        }
```
Will return a mask like this: `mask[name,id]`, which isn't going to work for getObject calls.


Basically, we should check to see if the objectMask starts with the call method name, and if so, drop the first bits of the objectMask, which I think can be done in https://github.com/softlayer/softlayer-java/blob/master/src/main/java/com/softlayer/api/RestApiClient.java#L455

basically if your doing  mask[blockDeviceTemplateGroups.id] and Account::getBlockDeviceTemplateGroups(), mask should be changed to mask[id], but if you are using Account::getObject() mask should be unchanged.

**Comment by FernandoOjeda** [comment]
The changes in the code were not to use getObject next to the required method to be able to use objectMask and resultLimit at the same time.

`List <Group> = accountService.getBlockDeviceTemplateGroups ()

resultLimit does not work using getObject () between accountService and getBlockDeviceTemplateGroups () like below, it does not recognize it.

`List <Group> = accountService.getObject.getBlockDeviceTemplateGroups ()

`I try the mask example below and it works fine without using getObject.

```
Account.Service accountService = Account.service (client);

        accountService.withMask (). abuseEmail ();
        accountService.withMask (). virtualGuests ();
        accountService.withMask (). blockDeviceTemplateGroups (). id ();
        accountService.withMask (). blockDeviceTemplateGroups (). name ();

        System.out.print ("mask [" + accountService.withMask (). ToString () + "] \ n");
        for (Group bd: accountService.getBlockDeviceTemplateGroups ()) {
            System.out.println ("BD:" + bd.getId () + "-" + bd.getName ());
        }

```
If we want to use Account :: getObject, the code that I implemented does not removes any value and returns the value that is being sent and the mask is correctly constructed and works.

I use the following code example and it works well:

```
Account.Service accountService = Account.service (client);

        accountService.withMask (). abuseEmail ();
        accountService.withMask (). virtualGuests ();
        accountService.withMask (). networkStorage ();

        try {
            Account result = accountService.getObject ();
            // Print response in JSON format
            Gson gson = new GsonBuilder (). SetPrettyPrinting (). Create ();
            System.out.println (gson.toJson (result));
        } catch (Exception e) {
            System.out.println ("Error:" + e);
        }

```

**Comment by allmightyspiff** [comment]
Having logic in the toString() method is going to cause a lot of headaches. 

Ideally the withMask() feature should be able to build a mask based on the mask supported by the call itself, but that is going to take quite a lot of work to make happen. @camporter  says he has some work in progress on this issue, so I'll likely get what hes started and try to finish it.

**Pull request closed without merging**